galley/pkg/config/processing/snapshotter/analyzingdistributor.go

```diff
@@ -272,8 +274,9 @@ func (c *context) ForEach(col collection.Name, fn analysis.IteratorFn) {
 func (c *context) Canceled() bool {
 	select {
 	case <-c.cancelCh:
-		return true
+		c.isCancel = true
```

Comment:
I am not familiar with how analysis work. But when `cancelCh` is closed, isn't always into this case?

Reply:
From above reference 3 source files, we can confirm that both AnalyzingDistributor instance's `cancelAnalysis` and the context instance's `cancelCh` will share the same channel. The purpose is designed for canceling the previous analysis process if there is another new analysis come.
However, according to the original [source code](https://github.com/istio/istio/blob/3e057db1331d14a4d276c8a15f3eb20c588daae9/galley/pkg/config/processing/snapshotter/analyzingdistributor.go#L165), this assertion is incorrect because `c.cancelCh` could have been changed from new channel to close channel, then to new channel again if there are several goroutines to do analysis work. This means that the analysis is canceled to be true and not really complete the analysis process. So I add a new properties for the context to record that context instance has been closed or not.

Reply:
Is it possible to reproduce the error case in a test? This would, I think, make the issue easier to grasp and help avoid future regressions.

Reply:
@sushicw Yes, I give the step list below